cmd/source_profile.go

```diff
@@ -86,58 +105,86 @@ type SourceProfileConnectionPostgreSQL struct {
+		// Set default port for postgresql, which rarely changes.
+		pg.port = "5432"
 	}
-	if pwd, ok := params["password"]; ok {
-		pg.pwd = pwd
+	if pg.pwd == "" {
+		pg.pwd = conversion.GetPassword()
 	}
-	return pg
+
+	return pg, nil
 }
 
 type SourceProfileConnectionDynamoDB struct {
-	awsAccessKeyID     string // Same as AWS_ACCESS_KEY_ID environment variable
-	awsSecretAccessKey string // Same as AWS_SECRET_ACCESS_KEY environment variable
-	awsRegion          string // Same as AWS_REGION environment variable
-	dydbEndpoint       string // Same as DYNAMODB_ENDPOINT_OVERRIDE environment variable
-	schemaSampleSize   int64  // Number of rows to use for inferring schema (default 100,000)
+	// We don't need any connection parameters here as SDK reads them from the env variables.
+	schemaSampleSize int64 // Number of rows to use for inferring schema (default 100,000)
 }
 
 func NewSourceProfileConnectionDynamoDB(params map[string]string) (SourceProfileConnectionDynamoDB, error) {
 	// TODO: Populate environment variables for DynamoDB if passed as flags.
 	dydb := SourceProfileConnectionDynamoDB{}
-	if awsAccessKeyID, ok := params["awsAccessKeyID"]; ok {
-		dydb.awsAccessKeyID = awsAccessKeyID
-	}
-	if awsSecretAccessKey, ok := params["awsSecretAccessKey"]; ok {
-		dydb.awsSecretAccessKey = awsSecretAccessKey
-	}
-	if awsRegion, ok := params["awsRegion"]; ok {
-		dydb.awsRegion = awsRegion
-	}
-	if dydbEndpoint, ok := params["dydbEndpoint"]; ok {
-		dydb.dydbEndpoint = dydbEndpoint
-	}
 	if schemaSampleSize, ok := params["schemaSampleSize"]; ok {
 		schemaSampleSizeInt, err := strconv.Atoi(schemaSampleSize)
 		if err != nil {
 			return dydb, fmt.Errorf("could not parse schemaSampleSize = %v as a valid int64", schemaSampleSize)
 		}
 		dydb.schemaSampleSize = int64(schemaSampleSizeInt)
 	}
+	awsAccessKeyID, idOk := params["awsAccessKeyID"]
+	awsSecretAccessKey, keyOk := params["awsSecretAccessKey"]
+	awsRegion, regionOk := params["awsRegion"]
+	dydbEndpoint, endpointOk := params["dydbEndpoint"]
+
+	if !(idOk || keyOk || regionOk || endpointOk) {
+		// No connection params provided through source-profile. Fetching from env variables.
+		fmt.Printf("Connection parameters not specified in source-profile. Reading from " +
+			"environment variables AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_REGION, DYNAMODB_ENDPOINT_OVERRIDE(optional)...\n")
+		// Don't need to do anything as they are handled later.
+	} else if idOk && keyOk && regionOk {
+		// Override env variables for these params as source-profile takes precedence.
+		os.Setenv("AWS_ACCESS_KEY_ID", awsAccessKeyID)
+		os.Setenv("AWS_SECRET_ACCESS_KEY", awsSecretAccessKey)
+		os.Setenv("AWS_REGION", awsRegion)
+		// Endpoint is optional. If not provided, the SDK infers endpoint from AWS_REGION.
+		// We should explicitly set it to "" if not provided to handle the case
+		// when the user already has an env variable DYNAMODB_ENDPOINT_OVERRIDE.
+		if !endpointOk {
+			dydbEndpoint = ""
+		}
```

Comment:
Is this the right thing to do? i.e., should we try to clear the endpoint without telling the user - I think that will get rather confusing. Updating the environment variables is probably okay but not clearing them IMO. If user did wanted to clear it, they would explicitly set it to empty string in source-profile.

Also, the case of DynamoDB is different from PostgreSQL and MySQL I think. There it was clear that we want to deprecate setting env variables. But for the case of DynamoDB it's perfectly alright to set through environment variables (and is the preferred way), and I wonder if it's better to rewrite this as following instead?

if dydb.awsAccessKeyID, ok := params["awsAccessKeyID"]; ok {
   os.Setenv("AWS_ACCESS_KEY_ID", awsAccessKeyID)
}
...

and so on.